File: csharp/SampleGameBackend/Controllers/AssetsController.cs
using Microsoft.AspNetCore.Mvc;
using SampleGameBackend.Database;
using SampleGameBackend.Models;
using SampleGameBackend.Middleware;

namespace SampleGameBackend.Controllers
{
    [ApiController]
    [Route("api")]
    public class AssetsController : ControllerBase
    {
        private readonly DatabaseService _databaseService;
        private readonly ILogger<AssetsController> _logger;

        public AssetsController(DatabaseService databaseService, ILogger<AssetsController> logger)
        {
            _databaseService = databaseService;
            _logger = logger;
        }

        [HttpGet("assets")]
        public IActionResult GetAssets([FromQuery] string? language)
        {
            try
            {
                // Validate session ID
                var sessionId = HttpContext.Items["SessionId"] as string;
                if (string.IsNullOrEmpty(sessionId))
                {
                    return BadRequest(new { success = false, errorCode = "INVALID_SESSION" });
                }

                // Validate language parameter
                if (string.IsNullOrEmpty(language))
                {
                    language = "ko"; // Default value
                }

                // Get or create session-specific asset information
                var sessionAssets = _databaseService.GetOrCreateSessionAssets(sessionId);

                // Convert to Asset struct
                var assets = new List<Asset>();
                foreach (var kvp in sessionAssets.Assets)
                {
                    assets.Add(new Asset
                    {
                        Id = kvp.Key,
                        Balance = kvp.Value
                    });
                }

                var v1Data = new V1Data
                {
                    PlayerId = sessionId,
                    Name = $"playerName_{sessionId}",
                    WalletAddress = "0xaaaa",
                    Server = "test",
                    Assets = assets
                };

                // Parse session information
                var createdAt = DateTime.Parse(sessionAssets.CreatedAt);
                var updatedAt = DateTime.Parse(sessionAssets.UpdatedAt);

                var guide = new
                {
                    Authorization = HttpContext.Request.Headers["Authorization"].FirstOrDefault(),
                    DappAuth = HttpContext.Request.Headers["X-Dapp-Authorization"].FirstOrDefault(),
                    SessionID = sessionId,
                    Message = "The guide field displays header information at request time. It is used to verify that the game company and protocol are correctly matched and is not provided to the game company. For ramp frontend developer reference.",
                    SessionInfo = new
                    {
                        CreatedAt = createdAt,
                        UpdatedAt = updatedAt
                    }
                };

                var response = new Response
                {
                    Success = true,
                    ErrorCode = null,
                    Data = new ResponseData
                    {
                        V1 = v1Data,
                        Guide = guide
                    }
                };

                return Ok(response);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "GetAssets error for sessionId: {SessionId}", 
                    HttpContext.Items["SessionId"] as string);
                return StatusCode(500, new { success = false, errorCode = "DB_ERROR" });
            }
        }
    }
} 

File: typescript/src/controllers/assetsController.ts
import { Response } from 'express';
import { AuthenticatedRequest } from '../middleware/authMiddleware';
import { AssetsResponse, Asset } from '../types';
import { MemoryDatabase } from '../database/memoryDb';

export class AssetsController {
  private db: MemoryDatabase;

  constructor(db: MemoryDatabase) {
    this.db = db;
  }

  /**
   * Get assets for a session
   * @param req - Request object
   * @param res - Response object
   */
  getAssets(req: AuthenticatedRequest, res: Response): void {
    try {
      const sessionId = req.sessionId;
      const language = (req.query.language as string) || 'ko';

      if (!sessionId) {
        res.status(400).json({
          success: false,
          errorCode: 'INVALID_SESSION_ID',
          message: 'Session ID is required'
        });
        return;
      }

      // Get session assets from database
      const sessionAssets = this.db.getSessionAssets(sessionId);
      
      // TODO: In actual implementation, fetch from real database
      // For demo purposes, create mock assets if not found
      const assets: Asset[] = sessionAssets?.assets || [
        { id: 'asset_money', balance: '2000' },
        { id: 'item_gem', balance: '1500' }
      ];

      // Create guide information
      const guide = {
        Authorization: `Bearer ${req.headers.authorization}`,
        'X-Dapp-Authorization': req.headers['x-dapp-authorization'] as string,
        'X-Dapp-SessionID': sessionId,
        message: 'guide 필드는 요청 시 헤더 정보를 표기합니다. 올바르게 게임사와 프로토콜을 맞췄는지 확인하는 용도이고 게임사에는 제공되지 않습니다. ramp frontend 개발자 참고용입니다.',
        session_info: {
          created_at: new Date().toISOString(),
          updated_at: new Date().toISOString()
        }
      };

      const response: AssetsResponse = {
        success: true,
        errorCode: null,
        data: {
          v1: {
            player_id: 'C1',
            name: 'playerName_C1',
            wallet_address: '0xaaaa',
            server: 'test',
            assets
          },
          guide
        }
      };

      res.json(response);
    } catch (error) {
      console.error('Error in getAssets:', error);
      res.status(500).json({
        success: false,
        errorCode: 'DB_ERROR',
        message: 'Internal server error'
      });
    }
  }
} 